enterprise/backend/src/metabase_enterprise/semantic_search/db/locking.clj

```diff
@@ -0,0 +1,57 @@
+(ns metabase-enterprise.semantic-search.db.locking
+  (:require
+   [honey.sql :as sql]
+   [honey.sql.helpers :as sql.helpers]
+   [metabase-enterprise.semantic-search.db.util :as semantic.db.util]
+   [metabase.util.log :as log]
+   [next.jdbc :as jdbc])
+  (:import
+   (java.sql SQLException)))
+
+(set! *warn-on-reflection* true)
+
+(def ^:private supported-lock-types
+  #{;; blocking exclusive transaction lock
+    :pg_advisory_xact_lock
+    ;; non-blocking exclusive transaction lock
+    :pg_try_advisory_xact_lock
+    ;; non-blocking shared transaction lock
+    :pg_try_advisory_xact_lock_shared})
+
+(defn- lock-not-avail-ex?
+  "SQL locking exceptino?"
+  [e]
+  (boolean
+   (when (instance? SQLException e)
+     (= "55P03" (.getSQLState ^SQLException e)))))
+
+(defn lock-or-throw!
```

Comment:
would be good to throw `ex-info` with `lock-type`/`lock-id` in the map instead of just Exception, in case we have more in the future